src/gui/share.cpp

```diff
@@ -0,0 +1,268 @@
+    job->getShares(path);
+}
+
+void ShareManager::slotSharesFetched(const QVariantMap &reply)
+{
+    QString message;
+    int code = OcsShareJob::getJsonReturnCode(reply, message);
+    if (code != 100 && code != 404) {
+        //emit error!
+    }
+
+    auto tmpShares = reply.value("ocs").toMap().value("data").toList();
+    const QString versionString = _account->serverVersion();
+    qDebug() << Q_FUNC_INFO << versionString << "Fetched" << tmpShares.count() << "shares";
+
+    QList<QSharedPointer<Share>> shares;
+
+    Q_FOREACH(auto share, tmpShares) {
+        auto data = share.toMap();
+
+        auto shareType = data.value("share_type").toInt();
+
+        Share *newShare = NULL;
+
+        if (shareType == OcsShareJob::ShareType::Link) {
+            newShare = parseLinkShare(data);
+        } else {
+            newShare = new Share(_account,
+                                 data.value("id").toString(),
+                                 data.value("path").toString(),
+                                 shareType,
+                                 data.value("permissions").toInt(),
+                                 this);
+        }
+
+        shares.append(QSharedPointer<Share>(newShare));    
+    }
+
+    qDebug() << Q_FUNC_INFO << "Sending " << shares.count() << "shares";
+    emit sharesFetched(shares);
+}
+
+LinkShare *ShareManager::parseLinkShare(const QVariantMap &data) {
+    QUrl url;
+    const QString versionString = _account->serverVersion();
+
+    // From ownCloud server 8.2 the url field is always set for public shares
+    if (data.contains("url")) {
+        url = QUrl(data.value("url").toString());
+    } else if (versionString.contains('.') && versionString.split('.')[0].toInt() >= 8) {
+        // From ownCloud server version 8 on, a different share link scheme is used.
+        url = QUrl(Account::concatUrlPath(_account->url(), QString("index.php/s/%1").arg(data.value("token").toString())).toString());
+    } else {
+        QList<QPair<QString, QString>> queryArgs;
+        queryArgs.append(qMakePair(QString("service"), QString("files")));
+        queryArgs.append(qMakePair(QString("t"), data.value("token").toString()));
+        url = QUrl(Account::concatUrlPath(_account->url(), QLatin1String("public.php"), queryArgs).toString());
+    }
+
+    QDate expireDate;
+    if (data.value("expiration").isValid()) {
+       expireDate = QDate::fromString(data.value("expiration").toString(), "yyyy-MM-dd 00:00:00");
+    }
+
+    return new LinkShare(_account,
+                         data.value("id").toString(),
+                         data.value("path").toString(),
+                         data.value("share_type").toInt(),
+                         data.value("permissions").toInt(),
+                         data.value("share_with").isValid(),
+                         url,
+                         expireDate,
+                         this);
```

Comment:
Hmm. You put pointers to these objects into QSharedPointer *and* parent them to this. Who owns these objects?